crates/re_viewer_context/src/collapsed_id.rs

```diff
@@ -0,0 +1,105 @@
+//! Helper types for producing stable [`egui::Id`] for the purpose of handling collapsed state of
+//! various UI elements.
+
+use crate::{ContainerId, SpaceViewId};
+use re_log_types::EntityPath;
+use std::marker::PhantomData;
+
+/// Defines a "scope" for the purpose of scoping stable [`egui::Id`]s.
+///
+/// Typically, if the same piece of data appears in two distinct UI areas, it's collapsed state
+/// shouldn't be shared. The generated [`egui::Id`] should thus be scoped to the actual UI tree
+/// involved. That can be done by defining a scope for each UI area, returning a unique identifier.
+pub trait CollapsedIdScope {
+    /// Unique identifier for the scope.
+    fn identifier() -> &'static str;
+}
+
+/// A collapsed identifier.
+///
+/// This enum helps to generate scoped, stable [`egui::Id`]s for various pieces of data.
+#[derive(Debug, Clone)]
+pub enum CollapsedId<Scope: crate::collapsed_id::CollapsedIdScope> {
```

Comment:
I'm curious why you went with a static type-system approach to the scope, instead of combining two enums in a `struct`. I don't care much each way, but using an enum would allow easier "uncollapse this everywhere" behavior, i.e.:

```rs
enum CollapseScope { Streams, Blueprint, Selection, … }
enum CollapseParticular { Container(…), SpaceView(…), … }
struct CollapseId { scope: CollapseScope, particular: CollapseParticular }

impl CollapseParticular {
    /// Open/collapse in all scopes
    pub  fn set_open_everywhere(&self, ctx: &egui::Context, open: bool) { 
        for scope in CollapseScope::ALL {
            CollapseId::new(scope, self.clone()).set_open(open);
        }
    }
}
```

Reply:
The honest answer to that is historical. Initially I had the scoped `type` definitions local to UI code (e.g  `BlueprintCollapseId` was in `viewport_blueprint_ui.rs`, etc.). I think my approach made sense in that context. But then I realised that I needed to make these scopes globally available for "remote control" of collapsedness. This is why I moved them back to `collapsed_id.rs`, but yeah, that design makes little sense now.